WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift

```diff
@@ -48,31 +49,28 @@ final class ProductFormViewController: UIViewController {
     private var cancellableProductName: ObservationToken?
     private var cancellableUpdateEnabled: ObservationToken?
 
-    init(product: Product,
+    init(viewModel: ViewModel,
+         productImageActionHandler: ProductImageActionHandler,
          currency: String = CurrencySettings.shared.symbol(from: CurrencySettings.shared.currencyCode),
-         presentationStyle: PresentationStyle,
+         presentationStyle: ProductFormPresentationStyle,
          isEditProductsRelease2Enabled: Bool,
          isEditProductsRelease3Enabled: Bool) {
+        self.viewModel = viewModel
         self.currency = currency
         self.presentationStyle = presentationStyle
         self.isEditProductsRelease2Enabled = isEditProductsRelease2Enabled
         self.isEditProductsRelease3Enabled = isEditProductsRelease3Enabled
-        self.productImageActionHandler = ProductImageActionHandler(siteID: product.siteID,
-                                                                   product: product)
+        self.productImageActionHandler = productImageActionHandler
         self.productUIImageLoader = DefaultProductUIImageLoader(productImageActionHandler: productImageActionHandler,
                                                                 phAssetImageLoaderProvider: { PHImageManager.default() })
-        self.viewModel = ProductFormViewModel(product: product,
-                                              productImageActionHandler: productImageActionHandler,
-                                              isEditProductsRelease2Enabled: isEditProductsRelease2Enabled,
-                                              isEditProductsRelease3Enabled: isEditProductsRelease3Enabled)
-        self.tableViewModel = DefaultProductFormTableViewModel(product: product,
+        self.tableViewModel = DefaultProductFormTableViewModel(product: viewModel.productModel,
                                                                actionsFactory: viewModel.actionsFactory,
                                                                currency: currency)
         self.tableViewDataSource = ProductFormTableViewDataSource(viewModel: tableViewModel,
                                                                   productImageStatuses: productImageActionHandler.productImageStatuses,
                                                                   productUIImageLoader: productUIImageLoader,
                                                                   canEditImages: isEditProductsRelease2Enabled)
-        super.init(nibName: nil, bundle: nil)
+        super.init(nibName: "ProductFormViewController", bundle: nil)
```

Comment:
[Q] Can we use String(describing: self) or it won't work due to generics?

Reply:
I can't use `String(describing: self)` since it's not after the self init (error: `'self' used before 'super.init' call`). I did try `type(of: self).nibName` and it didn't work because the type includes the generics like `ProductFormViewController<ProductFormViewModel>` and it crashes the app because it couldn't find a nib with this name that includes generics 😞